File: backend/scripts/provision.js
import { readFileSync, existsSync } from 'node:fs';
import { resolve } from 'node:path';

import {
  Client,
  Functions,
  Permission,
  Role,
  Storage,
  TablesDB,
} from 'node-appwrite';

loadDotEnv();

const { collectionIds, databaseId } = await import('../src/ids.js');
const { defaultCategories, defaultProducts, defaultUnits } = await import(
  '../src/seed-data.js'
);
const { appwriteSchema } = await import('../src/schema.js');
const { buildProductDocument } = await import('../src/rules.js');
const { nowIso } = await import('../src/normalization.js');

const client = new Client()
  .setEndpoint(requireEnv('APPWRITE_ENDPOINT'))
  .setProject(requireEnv('APPWRITE_PROJECT_ID'))
  .setKey(requireEnv('APPWRITE_API_KEY'));

const databases = new ProvisionTablesDB(new TablesDB(client));
const storage = new ProvisionStorage(new Storage(client));
const functions = new ProvisionFunctions(new Functions(client));

await provision();

async function provision() {
  await ensureDatabase();

  for (const collection of appwriteSchema.collections) {
    await ensureCollection(collection);
    await ensureAttributes(collection);
    await sleep(Number(process.env.MOONA_PROVISION_WAIT_MS || 750));
    await ensureIndexes(collection);
  }

  await ensureBucket();
  await seedCatalogs();
  await ensureFunctions();

  console.log('Moona Appwrite backend provisioning complete.');
}

async function ensureDatabase() {
  await ignoreConflict(() =>
    databases.create({
      databaseId: appwriteSchema.database.id,
      name: appwriteSchema.database.name,
    }),
  );
}

async function ensureCollection(collection) {
  await ignoreConflict(() =>
    databases.createCollection({
      databaseId,
      collectionId: collection.id,
      name: collection.name,
      permissions: collection.documentSecurity
        ? []
        : [Permission.read(Role.users())],
      documentSecurity: collection.documentSecurity,
      enabled: true,
    }),
  );
}

async function ensureAttributes(collection) {
  for (const attribute of collection.attributes) {
    await ignoreConflict(() => createAttribute(collection.id, attribute));
  }
}

async function createAttribute(collectionId, attribute) {
  const base = {
    databaseId,
    collectionId,
    key: attribute.key,
    required: attribute.required,
    array: Boolean(attribute.array),
  };

  if (attribute.type === 'string') {
    return databases.createStringAttribute({
      ...base,
      size: attribute.size,
      default: attribute.default,
    });
  }
  if (attribute.type === 'integer') {
    return databases.createIntegerAttribute({
      ...base,
      min: attribute.min,
      max: attribute.max,
      default: attribute.default,
    });
  }
  if (attribute.type === 'float') {
    return databases.createFloatAttribute({
      ...base,
      min: attribute.min,
      max: attribute.max,
      default: attribute.default,
    });
  }
  if (attribute.type === 'boolean') {
    return databases.createBooleanAttribute({
      ...base,
      default: attribute.default,
    });
  }
  if (attribute.type === 'datetime') {
    return databases.createDatetimeAttribute({
      ...base,
      default: attribute.default,
    });
  }
  if (attribute.type === 'enum') {
    return databases.createEnumAttribute({
      ...base,
      elements: attribute.elements,
      default: attribute.default,
    });
  }

  throw new Error(`Unsupported attribute type ${attribute.type}`);
}

async function ensureIndexes(collection) {
  for (const index of collection.indexes) {
    await ignoreConflict(() =>
      databases.createIndex({
        databaseId,
        collectionId: collection.id,
        key: index.key,
        type: index.type,
        attributes: index.attributes,
        orders: index.orders || [],
      }),
    );
  }
}

async function ensureBucket() {
  await ignoreConflict(() =>
    storage.createBucket({
      bucketId: appwriteSchema.bucket.id,
      name: appwriteSchema.bucket.name,
      permissions: [],
      fileSecurity: appwriteSchema.bucket.fileSecurity,
      enabled: appwriteSchema.bucket.enabled,
      maximumFileSize: appwriteSchema.bucket.maximumFileSize,
      allowedFileExtensions: appwriteSchema.bucket.allowedFileExtensions,
    }),
  );
}

async function seedCatalogs() {
  for (const category of defaultCategories) {
    await upsertDocument(collectionIds.categories, category.id, {
      stableId: category.id,
      nameAr: category.nameAr,
      nameEn: category.nameEn,
      emoji: category.emoji,
      sortOrder: category.sortOrder,
      active: category.active,
      createdAt: nowIso(),
      updatedAt: nowIso(),
    });
  }

  for (const unit of defaultUnits) {
    await upsertDocument(collectionIds.units, unit.id, {
      stableId: unit.id,
      nameAr: unit.nameAr,
      nameEn: unit.nameEn,
      sortOrder: unit.sortOrder,
      active: unit.active,
      createdAt: nowIso(),
      updatedAt: nowIso(),
    });
  }

  for (const product of defaultProducts) {
    const document = buildProductDocument(product, product.id);
    await upsertDocument(collectionIds.products, product.id, removeId(document));
  }
}

async function upsertDocument(collectionId, documentId, data) {
  await ignoreConflict(
    () =>
      databases.createDocument({
        databaseId,
        collectionId,
        documentId,
        data,
        permissions: [Permission.read(Role.users())],
      }),
    () =>
      databases.updateDocument({
        databaseId,
        collectionId,
        documentId,
        data: { ...withoutCreatedAt(data), updatedAt: nowIso() },
      }),
  );
}

async function ensureFunctions() {
  for (const spec of appwriteSchema.functions) {
    await ignoreConflict(() =>
      functions.create({
        functionId: spec.functionId,
        name: spec.name,
        runtime: spec.runtime,
        execute: spec.execute,
        events: [],
        schedule: '',
        timeout: spec.timeout,
        enabled: true,
        logging: true,
        entrypoint: spec.entrypoint,
        commands: spec.commands,
        scopes: [
          'databases.read',
          'tables.read',
          'rows.read',
          'rows.write',
          'users.read',
          'users.write',
          'buckets.read',
          'files.read',
          'files.write',
        ],
      }),
    );
    await ensureFunctionVariables(spec);
  }
}

async function ensureFunctionVariables(spec) {
  const variables = await functions.listVariables(spec.functionId);
  const existingByKey = new Map(
    variables.variables.map((variable) => [variable.key, variable]),
  );

  for (const [key, value] of Object.entries(functionVariables(spec))) {
    const existing = existingByKey.get(key);
    if (existing) {
      await functions.updateVariable({
        functionId: spec.functionId,
        variableId: existing.$id,
        key,
        value,
        secret: false,
      });
    } else {
      await functions.createVariable({
        functionId: spec.functionId,
        key,
        value,
        secret: false,
      });
    }
  }
}

function functionVariables(spec) {
  return {
    ...spec.environment,
    APPWRITE_ENDPOINT: requireEnv('APPWRITE_ENDPOINT'),
    APPWRITE_PROJECT_ID: requireEnv('APPWRITE_PROJECT_ID'),
    MOONA_ADMIN_USER_IDS: process.env.MOONA_ADMIN_USER_IDS || '',
    MOONA_DEFAULT_LANGUAGE: process.env.MOONA_DEFAULT_LANGUAGE || 'ar',
    MOONA_DEFAULT_THEME: process.env.MOONA_DEFAULT_THEME || 'light',
  };
}

async function ignoreConflict(createFn, updateFn = null) {
  try {
    return await createFn();
  } catch (error) {
    if (isConflict(error)) {
      if (updateFn) return updateFn();
      return null;
    }
    throw error;
  }
}

function isConflict(error) {
  return error?.code === 409 || String(error?.type || '').includes('already');
}

function removeId(document) {
  const { id, ...rest } = document;
  return rest;
}

function withoutCreatedAt(document) {
  const { createdAt, ...rest } = document;
  return rest;
}

function loadDotEnv() {
  const file = resolve(process.cwd(), '.env');
  if (!existsSync(file)) return;

  const content = readFileSync(file, 'utf8');
  for (const line of content.split(/\r?\n/)) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    const index = trimmed.indexOf('=');
    if (index === -1) continue;
    const key = trimmed.slice(0, index).trim();
    const value = trimmed.slice(index + 1).trim();
    process.env[key] ||= value;
  }
}

function requireEnv(key) {
  const value = process.env[key];
  if (!value) throw new Error(`Missing required environment variable ${key}`);
  return value;
}

function sleep(ms) {
  return new Promise((resolveSleep) => {
    setTimeout(resolveSleep, ms);
  });
}

class ProvisionTablesDB {
  constructor(tablesDB) {
    this.tablesDB = tablesDB;
  }

  create({ databaseId, name, enabled }) {
    return this.tablesDB.create({ databaseId, name, enabled });
  }

  createCollection({
    databaseId,
    collectionId,
    name,
    permissions,
    documentSecurity,
    enabled,
  }) {
    return this.tablesDB.createTable({
      databaseId,
      tableId: collectionId,
      name,
      permissions,
      rowSecurity: documentSecurity,
      enabled,
    });
  }

  createStringAttribute({
    databaseId,
    collectionId,
    key,
    size,
    required,
    default: defaultValue,
    array,
  }) {
    return this.tablesDB.createVarcharColumn({
      databaseId,
      tableId: collectionId,
      key,
      size,
      required,
      xdefault: defaultValue,
      array,
    });
  }

  createIntegerAttribute({
    databaseId,
    collectionId,
    key,
    required,
    min,
    max,
    default: defaultValue,
    array,
  }) {
    return this.tablesDB.createIntegerColumn({
      databaseId,
      tableId: collectionId,
      key,
      required,
      min,
      max,
      xdefault: defaultValue,
      array,
    });
  }

  createFloatAttribute({
    databaseId,
    collectionId,
    key,
    required,
    min,
    max,
    default: defaultValue,
    array,
  }) {
    return this.tablesDB.createFloatColumn({
      databaseId,
      tableId: collectionId,
      key,
      required,
      min,
      max,
      xdefault: defaultValue,
      array,
    });
  }

  createBooleanAttribute({
    databaseId,
    collectionId,
    key,
    required,
    default: defaultValue,
    array,
  }) {
    return this.tablesDB.createBooleanColumn({
      databaseId,
      tableId: collectionId,
      key,
      required,
      xdefault: defaultValue,
      array,
    });
  }

  createDatetimeAttribute({
    databaseId,
    collectionId,
    key,
    required,
    default: defaultValue,
    array,
  }) {
    return this.tablesDB.createDatetimeColumn({
      databaseId,
      tableId: collectionId,
      key,
      required,
      xdefault: defaultValue,
      array,
    });
  }

  createEnumAttribute({
    databaseId,
    collectionId,
    key,
    elements,
    required,
    default: defaultValue,
    array,
  }) {
    return this.tablesDB.createEnumColumn({
      databaseId,
      tableId: collectionId,
      key,
      elements,
      required,
      xdefault: defaultValue,
      array,
    });
  }

  createIndex({ databaseId, collectionId, key, type, attributes, orders }) {
    return this.tablesDB.createIndex({
      databaseId,
      tableId: collectionId,
      key,
      type,
      columns: attributes,
      orders,
    });
  }

  createDocument({ databaseId, collectionId, documentId, data, permissions }) {
    return this.tablesDB.createRow({
      databaseId,
      tableId: collectionId,
      rowId: documentId,
      data,
      permissions,
    });
  }

  updateDocument({ databaseId, collectionId, documentId, data, permissions }) {
    return this.tablesDB.updateRow({
      databaseId,
      tableId: collectionId,
      rowId: documentId,
      data,
      permissions,
    });
  }
}

class ProvisionStorage {
  constructor(storage) {
    this.storage = storage;
  }

  createBucket({
    bucketId,
    name,
    permissions,
    fileSecurity,
    enabled,
    maximumFileSize,
    allowedFileExtensions,
  }) {
    return this.storage.createBucket({
      bucketId,
      name,
      permissions,
      fileSecurity,
      enabled,
      maximumFileSize,
      allowedFileExtensions,
    });
  }
}

class ProvisionFunctions {
  constructor(functions) {
    this.functions = functions;
  }

  create({
    functionId,
    name,
    runtime,
    execute,
    events,
    schedule,
    timeout,
    enabled,
    logging,
    entrypoint,
    commands,
    scopes,
  }) {
    return this.functions.create({
      functionId,
      name,
      runtime,
      execute,
      events,
      schedule,
      timeout,
      enabled,
      logging,
      entrypoint,
      commands,
      scopes,
    });
  }

  listVariables(functionId) {
    return this.functions.listVariables({ functionId });
  }

  createVariable({ functionId, key, value, secret }) {
    return this.functions.createVariable({
      functionId,
      variableId: key,
      key,
      value,
      secret,
    });
  }

  updateVariable({ functionId, variableId, key, value, secret }) {
    return this.functions.updateVariable({
      functionId,
      variableId,
      key,
      value,
      secret,
    });
  }
}

File: backend/bin/provision.dart
import 'dart:io';

import 'package:dart_appwrite/dart_appwrite.dart';
import 'package:dart_appwrite/enums.dart';
import 'package:moona_backend/moona_backend.dart';

final _dotenv = <String, String>{};

Future<void> main() async {
  loadDotEnv();

  final client = Client()
      .setEndpoint(requireEnvValue('APPWRITE_ENDPOINT'))
      .setProject(requireEnvValue('APPWRITE_PROJECT_ID'))
      .setKey(requireEnvValue('APPWRITE_API_KEY'));

  final tablesDB = TablesDB(client);
  final storage = Storage(client);
  final functions = Functions(client);

  await provision(tablesDB, storage, functions);
  stdout.writeln('Moona Appwrite backend provisioning complete.');
}

Future<void> provision(
  TablesDB tablesDB,
  Storage storage,
  Functions functions,
) async {
  await ensureDatabase(tablesDB);

  for (final table in appwriteSchema.tables) {
    await ensureTable(tablesDB, table);
    await ensureAttributes(tablesDB, table);
    await Future<void>.delayed(
      Duration(milliseconds: intFromEnv('MOONA_PROVISION_WAIT_MS', 750)),
    );
    await ensureIndexes(tablesDB, table);
  }

  await ensureBucket(storage);
  await seedCatalogs(tablesDB);
  await ensureFunctions(functions);
}

Future<void> ensureDatabase(TablesDB tablesDB) async {
  try {
    await ignoreConflict(
      () => tablesDB.create(
        databaseId: appwriteSchema.databaseId,
        name: appwriteSchema.databaseName,
      ),
    );
  } catch (error) {
    if (!isAdditionalResourceLimit(error)) rethrow;

    final existing = await tablesDB.get(databaseId: appwriteSchema.databaseId);
    if (existing.$id != appwriteSchema.databaseId) rethrow;
  }
}

Future<void> ensureTable(TablesDB tablesDB, TableSpec table) async {
  await ignoreConflict(
    () => tablesDB.createTable(
      databaseId: databaseId,
      tableId: table.id,
      name: table.name,
      permissions: table.rowSecurity ? [] : [Permission.read(Role.users())],
      rowSecurity: table.rowSecurity,
      enabled: true,
    ),
  );
}

Future<void> ensureAttributes(TablesDB tablesDB, TableSpec table) async {
  for (final attribute in table.attributes) {
    await ignoreConflict(() => createAttribute(tablesDB, table.id, attribute));
  }
}

Future<dynamic> createAttribute(
  TablesDB tablesDB,
  String tableId,
  AttributeSpec attribute,
) {
  switch (attribute.type) {
    case 'string':
      return tablesDB.createVarcharColumn(
        databaseId: databaseId,
        tableId: tableId,
        key: attribute.key,
        size: attribute.size!,
        xrequired: attribute.required,
        xdefault: attribute.defaultValue as String?,
        array: attribute.array,
      );
    case 'integer':
      return tablesDB.createIntegerColumn(
        databaseId: databaseId,
        tableId: tableId,
        key: attribute.key,
        xrequired: attribute.required,
        min: attribute.min?.toInt(),
        max: attribute.max?.toInt(),
        xdefault: attribute.defaultValue as int?,
        array: attribute.array,
      );
    case 'float':
      return tablesDB.createFloatColumn(
        databaseId: databaseId,
        tableId: tableId,
        key: attribute.key,
        xrequired: attribute.required,
        min: attribute.min?.toDouble(),
        max: attribute.max?.toDouble(),
        xdefault: attribute.defaultValue as double?,
        array: attribute.array,
      );
    case 'boolean':
      return tablesDB.createBooleanColumn(
        databaseId: databaseId,
        tableId: tableId,
        key: attribute.key,
        xrequired: attribute.required,
        xdefault: attribute.defaultValue as bool?,
        array: attribute.array,
      );
    case 'datetime':
      return tablesDB.createDatetimeColumn(
        databaseId: databaseId,
        tableId: tableId,
        key: attribute.key,
        xrequired: attribute.required,
        xdefault: attribute.defaultValue as String?,
        array: attribute.array,
      );
    case 'enum':
      return tablesDB.createEnumColumn(
        databaseId: databaseId,
        tableId: tableId,
        key: attribute.key,
        elements: attribute.elements,
        xrequired: attribute.required,
        xdefault: attribute.defaultValue as String?,
        array: attribute.array,
      );
  }

  throw StateError('Unsupported attribute type ${attribute.type}');
}

Future<void> ensureIndexes(TablesDB tablesDB, TableSpec table) async {
  for (final index in table.indexes) {
    await ignoreConflict(
      () => tablesDB.createIndex(
        databaseId: databaseId,
        tableId: table.id,
        key: index.key,
        type: indexType(index.type),
        columns: index.attributes,
        orders:
            index.orders.isEmpty ? null : index.orders.map(orderBy).toList(),
      ),
    );
  }
}

Future<void> ensureBucket(Storage storage) async {
  final bucket = appwriteSchema.bucket;
  try {
    await ignoreConflict(
      () => storage.createBucket(
        bucketId: bucket.id,
        name: bucket.name,
        permissions: [],
        fileSecurity: bucket.fileSecurity,
        enabled: bucket.enabled,
        maximumFileSize: bucket.maximumFileSize,
        allowedFileExtensions: bucket.allowedFileExtensions,
      ),
    );
  } catch (error) {
    if (!isAdditionalResourceLimit(error)) rethrow;

    final existing = await storage.getBucket(bucketId: bucket.id);
    if (existing.$id != bucket.id) rethrow;
  }
}

Future<void> seedCatalogs(TablesDB tablesDB) async {
  for (final category in defaultCategories) {
    await upsertRow(
        tablesDB, CollectionIds.categories, category['id'] as String, {
      'stableId': category['id'],
      'nameAr': category['nameAr'],
      'nameEn': category['nameEn'],
      'emoji': category['emoji'],
      'sortOrder': category['sortOrder'],
      'active': category['active'],
      'createdAt': nowIso(),
      'updatedAt': nowIso(),
    });
  }

  for (final unit in defaultUnits) {
    await upsertRow(tablesDB, CollectionIds.units, unit['id'] as String, {
      'stableId': unit['id'],
      'nameAr': unit['nameAr'],
      'nameEn': unit['nameEn'],
      'sortOrder': unit['sortOrder'],
      'active': unit['active'],
      'createdAt': nowIso(),
      'updatedAt': nowIso(),
    });
  }

  for (final product in defaultProducts) {
    final document = buildProductDocument(product, product['id'] as String);
    await upsertRow(
      tablesDB,
      CollectionIds.products,
      product['id'] as String,
      removeId(document),
    );
  }
}

Future<void> upsertRow(
  TablesDB tablesDB,
  String tableId,
  String rowId,
  Map<String, dynamic> data,
) async {
  await ignoreConflict(
    () => tablesDB.createRow(
      databaseId: databaseId,
      tableId: tableId,
      rowId: rowId,
      data: data,
      permissions: [Permission.read(Role.users())],
    ),
    update: () => tablesDB.updateRow(
      databaseId: databaseId,
      tableId: tableId,
      rowId: rowId,
      data: {...withoutCreatedAt(data), 'updatedAt': nowIso()},
    ),
  );
}

Future<void> ensureFunctions(Functions functions) async {
  for (final spec in appwriteSchema.functions) {
    await ignoreConflict(
      () => functions.create(
        functionId: spec.functionId,
        name: spec.name,
        runtime: runtimeFrom(spec.runtime),
        execute: spec.execute,
        events: [],
        schedule: '',
        timeout: spec.timeout,
        enabled: true,
        logging: true,
        entrypoint: spec.entrypoint,
        commands: spec.commands,
        scopes: functionScopes,
      ),
      update: () => functions.update(
        functionId: spec.functionId,
        name: spec.name,
        runtime: runtimeFrom(spec.runtime),
        execute: spec.execute,
        events: [],
        schedule: '',
        timeout: spec.timeout,
        enabled: true,
        logging: true,
        entrypoint: spec.entrypoint,
        commands: spec.commands,
        scopes: functionScopes,
      ),
    );
    await ensureFunctionVariables(functions, spec);
  }
}

Future<void> ensureFunctionVariables(
  Functions functions,
  FunctionSpec spec,
) async {
  final variables = await functions.listVariables(functionId: spec.functionId);
  final existingByKey = <String, dynamic>{
    for (final variable in variables.variables) variable.key: variable,
  };

  for (final entry in functionVariables(spec).entries) {
    final existing = existingByKey[entry.key];
    if (existing == null) {
      await functions.createVariable(
        functionId: spec.functionId,
        variableId: entry.key,
        key: entry.key,
        value: entry.value,
        secret: false,
      );
    } else {
      await functions.updateVariable(
        functionId: spec.functionId,
        variableId: existing.$id.toString(),
        key: entry.key,
        value: entry.value,
        secret: false,
      );
    }
  }
}

Map<String, String> functionVariables(FunctionSpec spec) => {
      ...spec.environment,
      'APPWRITE_ENDPOINT': requireEnvValue('APPWRITE_ENDPOINT'),
      'APPWRITE_PROJECT_ID': requireEnvValue('APPWRITE_PROJECT_ID'),
      'MOONA_ADMIN_USER_IDS': envValue('MOONA_ADMIN_USER_IDS') ?? '',
      'MOONA_DEFAULT_LANGUAGE': envValue('MOONA_DEFAULT_LANGUAGE') ?? 'ar',
      'MOONA_DEFAULT_THEME': envValue('MOONA_DEFAULT_THEME') ?? 'light',
    };

Future<dynamic> ignoreConflict(
  Future<dynamic> Function() create, {
  Future<dynamic> Function()? update,
}) async {
  try {
    return await create();
  } catch (error) {
    if (isConflict(error)) {
      if (update != null) return update();
      return null;
    }
    rethrow;
  }
}

bool isConflict(Object error) {
  if (error is AppwriteException) {
    return error.code == 409 || (error.type ?? '').contains('already');
  }
  return false;
}

bool isAdditionalResourceLimit(Object error) {
  if (error is AppwriteException) {
    return error.code == 403 &&
        (error.type ?? '') == 'additional_resource_not_allowed';
  }
  return false;
}

Map<String, dynamic> removeId(Map<String, dynamic> document) {
  final copy = {...document};
  copy.remove('id');
  return copy;
}

Map<String, dynamic> withoutCreatedAt(Map<String, dynamic> document) {
  final copy = {...document};
  copy.remove('createdAt');
  return copy;
}

void loadDotEnv() {
  final file = File('.env');
  if (!file.existsSync()) return;

  for (final line in file.readAsLinesSync()) {
    final trimmed = line.trim();
    if (trimmed.isEmpty || trimmed.startsWith('#')) continue;
    final separator = trimmed.indexOf('=');
    if (separator <= 0) continue;
    final key = trimmed.substring(0, separator).trim();
    var value = trimmed.substring(separator + 1).trim();
    if ((value.startsWith('"') && value.endsWith('"')) ||
        (value.startsWith("'") && value.endsWith("'"))) {
      value = value.substring(1, value.length - 1);
    }
    _dotenv[key] = value;
  }
}

String requireEnvValue(String key) {
  final value = envValue(key);
  if (value == null || value.isEmpty) {
    throw StateError('Missing required environment variable $key');
  }
  return value;
}

String? envValue(String key) => _dotenv[key] ?? Platform.environment[key];

int intFromEnv(String key, int fallback) =>
    int.tryParse(envValue(key) ?? '') ?? fallback;

Runtime runtimeFrom(String runtime) {
  if (runtime == 'dart-3.1') return Runtime.dart31;
  if (runtime == 'dart-3.5') return Runtime.dart35;
  if (runtime == 'dart-3.8') return Runtime.dart38;
  if (runtime == 'dart-3.9') return Runtime.dart39;
  if (runtime == 'dart-3.10') return Runtime.dart310;
  if (runtime == 'dart-3.11') return Runtime.dart311;
  throw StateError('Unsupported function runtime $runtime');
}

TablesDBIndexType indexType(String type) {
  if (type == 'key') return TablesDBIndexType.key;
  if (type == 'unique') return TablesDBIndexType.unique;
  if (type == 'fulltext') return TablesDBIndexType.fulltext;
  throw StateError('Unsupported index type $type');
}

OrderBy orderBy(String order) {
  if (order == 'asc') return OrderBy.asc;
  if (order == 'desc') return OrderBy.desc;
  throw StateError('Unsupported index order $order');
}

const functionScopes = [
  ProjectKeyScopes.databasesRead,
  ProjectKeyScopes.tablesRead,
  ProjectKeyScopes.rowsRead,
  ProjectKeyScopes.rowsWrite,
  ProjectKeyScopes.usersRead,
  ProjectKeyScopes.usersWrite,
  ProjectKeyScopes.bucketsRead,
  ProjectKeyScopes.filesRead,
  ProjectKeyScopes.filesWrite,
  ProjectKeyScopes.tokensWrite,
  ProjectKeyScopes.messagesWrite,
];
